### ISSUE TITLE
Issues uploaded CTT data to Motus

### ISSUE THREAD

**Issue by leberrigan** [MotusDev/Motus-TO-DO#1038]
Someone uploaded data from their SensorStation which only contained CTT (434 MHz) detections. The job failed because SGData  is treating it like SG data, not CTT data ('archives are corrupt'). See: https://sgdata.motus.org/status?jobID=9786963

**Comment by denislepage** [comment]
I don't think this is what it means. These 41 files have a name that match the SG detections as captured by a CTT station. I spot checked a couple in the original zip file, and they appear to indeed by corrupted. It's a small number relative to the entire file (3,300 gzip in that archive, 42 of them broken), but those are pulses data from the SG side.

There appears to be no detection found for this batch, but I am almost certain it's not related to these broken files, and I am not seeing anything that suggests that this archive contains CTT tag data. The file name pattern for CTT tags is different, and when tag finder encounters them, it puts them in a separate folder, and those files then get pushed to the AWS account for CTT to handle.

**Comment by StuMackenzie** [comment]
Should be noted that this file should contain CTT tags. It was a deployment site for numerous tags, or at least close enough that it should have a lot of CTT data.

**Comment by Hussell** [comment]
Denis is correct, those 41 files were ignored but the other 3,154 files were processed without error. No Lotek tags detected, but that's expected, right?

**Comment by denislepage** [comment]
For reference, this is receiver CTT-867459049273626. The user has uploaded 2 files, one containing sg file, the other ctt files. The archive for the job listed above contains almost exclusive (only?) GPS records and no pulses. They may not have plugged a SG antenna at all. Not sure why it still generated sg files, but all of them are very small.

The CTT file (called 13098_2021-10-12T14-34-44.378_ctt-data-rotated.2021-10-08_191005.zip) is in the sgm/upload folder, but I am not seeing a batch associated with it, and the files it contains do not appear in the CTT outgoing_archive where I would expect them. From the looks of it, motusServer did not run as intended on this archive (it should have simply moved the files to a folder called ctt/outgoing, which I monitoring to push to AWS.

@Hussell could you try to find that job? Maybe I just missed it.

**Comment by denislepage** [comment]
(edited)

For reference, this is receiver CTT-867459049273626. The user has uploaded 2 files, one containing sg file, the other ctt files. The archive for the job listed above contains almost exclusive (only?) GPS records and no pulses. They may not have plugged a SG antenna at all. Not sure why it still generated sg files, but all of them are very small. In any case, as Jeremy suggests, there should be no detections on the SG job, since there are no pulses.

The CTT file (called 13098_2021-10-12T14-34-44.378_ctt-data-rotated.2021-10-08_191005.zip), matches the Job ID = 9786962 (I had to remember to search by User ID, not receiver ID for upload jobs). It says that it "Moved 6306 CTT files for separate processing". Those are going to sgm/ctt_outgoing, where one of the queue procs will normally grab them and push to AWS. I'm currently not able to access that directory. From sgdata, I can get the unsorted head (ls -U | head) quickly, but anything else is stalling (e.g.  ls -U | wc -l). I wonder if the directory has become too full, and the uploads are potentially timing out.

**Comment by denislepage** [comment]
There are at least 100,000 files, perhaps much more, so that may explain the timeouts. We should modify motusServer to place these files in subdirectories (e.g. by upload date), which might help. Java relies on DirectoryStream to try and iterate through the directory, but it somehow never manages to get started. Each of the procs queue jobs eventually gets interrupted when the server restarts, since it has no timeout. Once in a while, the job reports one of those 2 errors:

```
java.io.UncheckedIOException: java.nio.file.FileSystemException: \\10.100.100.9\sgdata\sgm\ctt_outgoing: The specified network name is no longer available.  
java.io.UncheckedIOException: java.nio.file.FileSystemException: \\10.100.100.9\sgdata\sgm\ctt_outgoing: The semaphore timeout period has expired.  
```

**Comment by denislepage** [comment]
@Hussell  I am stoping motusServer for now. I tried to upload a small zip with 2 ctt files, but that job has been stuck for a while. Once the other normal jobs are finished, I may try to restart the NAS drive.

**Comment by denislepage** [comment]
@Hussell  OK, I manually forced the job related to my upload, and the others are now terminated. I am trying to restart the NAS, but it is taking its sweet time. Don't restart motusServer before I confirm. I should be able to restart it myself, assuming the NAS drive comes back online.

**Comment by denislepage** [comment]
Quick update. There are still issues accessing the folder where ctt files are placed for uploading to AWS. I suspect there is an issue with the disks, and I asked Andrew to look into it. 

After quite a bit of fiddling today, I tried several options, and eventually tried to restart the whole server and processes. I believe things are back online, but the status server is still broken. I am also still unable to access the ctt_outgoing folder, so we may be mostly back where we started, but I can't confirm the jobs are runnung again.

Bottomline, there is a substantial backlog of CTT data that remains unprocessed. We'll need to make some changes to the way this is being handled. Perhaps creating subfolders in the ctt_outgoing folder (from motusServer) may help, but that will require a few other changes.

Arg.

**Comment by Hussell** [comment]
The status server has been restarted and seems to be running normally now.

**Comment by Hussell** [comment]
Running "ls -f | wc -l" in /sgm/ctt_outgoing returned "885259" after 4m 37s.

**Comment by denislepage** [comment]
I received a notification late last night, after I finally gave up, indicating that the NAS had run a monthly test on its drive. That may have explained some of the slow down issues.

There may still be a disk problem, but perhaps not.

The other thing I found is that Windows may not have proper access to that shared folder. I asked Andrew to look into it. That may be the core of the problem.

Thanks for getting this back on track. Weird that I tried to get it restarted several times without success

**Comment by StuMackenzie** [comment]
For reference.

3 more stations that might be effected. Same project.

5853    CTT-867459049273493    CTT SensorStation    2019-07-19 - ?    2021-10-31 14:30    Bernheim Fire Tower    

5879    CTT-867459049274186    CTT SensorStation    2019-07-20 - ?    2021-10-27 11:28    Bernheim Education Center    

5883    CTT-867459049273626    CTT SensorStation    2019-07-20 - ?    2021-10-08 17:42    Bean House

**Comment by Hussell** [comment]
That's all the receivers for project 292. All their antennas are listed as 434MHz, so all their data should be sent to CTT, not processed by our tag finder, correct?

**Comment by StuMackenzie** [comment]
‘Shouldn’t’, especially since they are modem connected so we shouldn’t have to touch that data.

**Comment by StuMackenzie** [comment]
Related, but maybe deserves a separate issue - https://motus.org/data/receiver?id=4327 has been connected for a week or so, but no data flowing through Motus. @Hussell @adamdsmith

**Comment by leberrigan** [comment]
@denislepage @Hussell  -- Any updates? We still don't have any tag hits in our DB from these receivers.

**Comment by leberrigan** [comment]
Just got another collaborator with the same sort of issue with 2 SensorStations (manual upload of CTT data thru Motus):

>...the processing system returned a message “I don't know how to handle these files:” for the full list of data files. (this is job 10507921)
> The files  have a name structure like: CTT-A46E6F0D83E8-log.2021-04-23_150833.csv.gz and are contained in a single .7z compressed file.

**Comment by Hussell** [comment]
https://sgdata.motus.org/status?jobID=10507921

Same as earlier, none of those are SensorGnome data files so it was pointless to upload them directly to Motus. Either they need to upload SensorGnome files, or upload CTT files to CTT. Is this misunderstanding happening more frequently now for some reason?

**Comment by denislepage** [comment]
Not sure what you mean.

We have process to theoretically handle CTT files uploaded to motus that parse them into an upload folder based on the file name. (then a separate process was supposed to push them to AWS, though I don’t think the permission issue we discovered recently has been resolved yet).

How is this causing this error?

**Comment by denislepage** [comment]
Hm. Log seems to be a new file type that we should add to the parsing regex.

I think that’s the issue.

I just added log to motusConstants.R, under MOTUS_CTT_SS_DATAFILE_REGEX

I should now be deployed on sgdata.

If they are introducing a new file type, I’ll also need to check if it needs to be incorporated into our database during the import of the files we get from CTT.

**Comment by Hussell** [comment]
Sorry for my misunderstanding above.

Alright, the log says the 3539 CTT files which were recognized as CTT files were sent off to CTT, as planned. The log files don't seem to contain anything we're interested in, but we should probably send them on to CTT in case they need them. There's also a type of file labelled "telemetery", same deal. The list of types is getting rather unwieldy, does it make sense to just look for a CTT serial number followed by "-[a-z-]+\\.csv\\.gz"?

I don't see any activity corresponding to the file dates on the receiver's timeline, so there's another problem beyond the log files not being recognized.
https://motus.org/data/receiver/timeline?year=2021&id=4434

There are some uncompressed data files in the archive (e.g. "CTT-A46E6F0D83E8-raw-data.csv" instead of "CTT-A46E6F0D83E8-raw-data.csv.gz"). It looks like those are being ignored as well. "\\.csv(?:\\.gz)?"

I'm searching for old upload jobs which where similar files were ignored. I can rerun some of them to test the fixes, and rerun the rest so no-one has to re-upload anything.

Denis, are you working on the master branch of motusServer?

**Comment by denislepage** [comment]
If we’re going to do that, you probably need to escape the hyphen in the brackets.

"-[a-z\-]+\.csv\.gz"

Yes, we should also be able to deal with txt files outside of gz.

I accidentally pushed my changes to the Lotek6M branch, so we should repeat them in master. Are you handling that?

I am also trying to restart the motus sandbox server for dealing with CTT uploads, but it’s not restarting well at the moment. I’ll give it a bit more time.

**Comment by Hussell** [comment]
Yes, I'll handle the motusServer changes and deploying to sgdata.

**Comment by denislepage** [comment]
Permission issues seem partly resolved. The pushing of CTT files to AWS has now started.

There are nearly 1 million files, so it’ll take a little while, and CTT may not yet have hooked up the processing on their end, so no timeline yet when we’ll see the data in the database. I’ll contact Bob and Jessica to let them know.

**Comment by Hussell** [comment]
Pushing the previously overlooked files to CTT is proceeding smoothly. Should be done in a few hours.

I'm seeing multiple copies of various processes in the process queue, e.g. three copies of UploadCTTDataToS3 in motus_sandbox, three copies of ReadCTTFiles in motus. Is there any way for me to tell whether those are actually still running or not, besides waiting to see whether the database locks up or not?

**Comment by leberrigan** [comment]
I'm still not seeing any new data where I expect some. Do we need to re-upload or are we still waiting on processing?

**Comment by Hussell** [comment]
So far as I'm aware, we're only waiting for CTT. Denis might know more.

**Comment by Hussell** [comment]
Denis said on this week's call (heavily summarized): files still aren't making it to CTT. Working on it.

**Comment by denislepage** [comment]
So, I have been working through the backlog. It's slow going. I probably have now about a quarter of it uploaded, but most of those so far are only the dongle data on SG units. Almost all of them so far are also empty (size = 0) or only contain only a single row with the path of the file name, so I am not sure what that means. I suspect that the SG patch is creating new files regularly, whether there are detections of CTT tags or not. At least a few files have CTT tag data. 

I can't tell yet whether this is a bug that makes the files lose data, or if this only happens when there are no data anyway. We'll need following up on that.

On the plus side, I have been working on implementing a parallel thread processing of uploads and downloads from AWS, which seems to make that step a LOT more efficient. There may be other bottlenecks that may make it worthwhile to package the files into archives rather than try to transfer them individually (most of them have less than 1 kb of data), but there is an appreciable lag in establishing the individual connections. Parallelism seems to mostly address that, so we'll see.

I don't think I've come across SensorStation files yet in the ones I have processed. More to come.

**Comment by Hussell** [comment]
They symlinks bug is fixed.

The bug was in the R code to detect the presence of a symlink. It was:

`targ = Sys.readlink(src)`
`iTarg = which(isTRUE(nchar(targ) > 0)) ## only files which are valid symlinks pass this test`
`if(length(iTarg)) {`

The bug is one which an experienced R programmer (so, not me) would likely spot immediately: isTRUE. All the other code in there works with a list, but that function only returns TRUE if it's passed a single TRUE. So the code only worked when there was one file (as in the by-hand tests I was doing earlier), but failed when there were multiple files. The worst part is that isTRUE is totally superfluous. The correct code is:

`targ = Sys.readlink(src)`
`iTarg = which(nchar(targ) > 0) ## only files which are valid symlinks pass this test`
`if(length(iTarg)) {`

**Comment by denislepage** [comment]
I have been monitoring the ctt_outgoing folder today. So far so good. I pushed a bunch of new files to AWS, and cleared any remaining symlinks in that folder. So far, none have reappeared. Woohoo.

A few left for me to do before we can close this: 

1. see if any of the files already moved to AWS keep reappearing. There may be some remnant files in file_repo that will get moved again one last time, but so far, I am not seeing the flurry of older files like last time.

2. I can switch back the location of the ctt_outgoing folder to the one where sgdata is moving those files, rather than a secondary drive.

3. We'll then need to confirm that the data are making their way back to us from CTT once they confirm that they have these hooked up. I expect that SensorStations files will be first, and then SG dongle data. That could take a few weeks to get there.

**Comment by ajmacdonald3** [comment]
Has it been confirmed that the data from dual mode manual uploads are making it back to us from CTT? If manually uploading data from a SensorStation, the folder structure within the zipped folder that gets uploaded to Motus should not matter and can contain some folders with SG/166 MHz data and some folders with 434 MHz data because tagfinder uses the individual file names to identify which ones to pull out and send to CTT, and then those will later be sent back to Motus. Am I correct with that interpretation?

**Comment by denislepage** [comment]
Interpretation is correct, but I haven’t heard a confirmation from CTT yet that they started processing the files. They now have over a million files to process in their hands for the dongle data.  I can check with Bob.

The files for CTT tags that are manually uploaded should also be making their way to them, after we solved our upload issues last month, and I expect those will show up first, if they are not already there.

Files that are uploaded to us via Motus and that contain CTT detections (whether collected on SG or SS) go to Amazon AWS first, CTT processes them and filter out the false positives, and then push them back to us. The process probably should/could be improved.

**Comment by leberrigan** [comment]
Just looking for an update on this - do you know if manual uploads are getting processed yet?

**Comment by denislepage** [comment]
CTT said today that they are now looking into those files.

**Comment by leberrigan** [comment]
CTT has confirm that it is working. Someone from our team should confirm that it's working on our end too.

**Comment by j-sayers** [comment]
I [had this SS](https://motus.org/data/receiverDeployment?id=8516) on for a few days around some test tags, with the cell option disabled.

I downloaded the CTT and SG data from the receiver and uploaded the two zips exactly as they were downloaded from the receiver.

Upload job ID 14322498 (CTT data)
Upload job ID 14322539 (SG data)

**Comment by denislepage** [comment]
I am seeing 2 batches for this receiver created today at 9:59 this morning, so about 45 minutes after the previous git post. Both seem to refer to CTT batches, one has 0 hits, and the other 87.

Fri Mar 25 2022 17:20:12 GMT+0000 to Sat Mar 26 2022 07:20:11 GMT+0000 (0 hits)
Sat Mar 26 2022 07:30:51 GMT+0000 to Mon Mar 28 2022 18:30:51 GMT+0000 (87 hits)

I have not yet seen anything related to Lotek data. The sgdata status server is returning no jobs. It is running @Hussell ?

**Comment by denislepage** [comment]
Sorry, got it backward. The 2 existing batches ARE for Lotek data. We haven't seen CTT tags come back yet that I know of.

**Comment by j-sayers** [comment]
The processing of CTT files job was this https://sgdata.motus.org/status?jobID=14322498

>         Logs by subjob:
> Log for 14322498
> 
> Job 14322498 with 1 pending subjobs entered processing queue 4
> 
> Log for 14324641
> 
> Moved 219 CTT files for separate processing with names like
>    /sgm/queue/4//14322498/14324647/data/rotated/CTT-0825A4690059-gps.2022-03-25_182041.csv.gz
> 
> Log for 14324647
> 
> Unpacking file 19829_2022-04-05T13-10-06.577_ctt-data-rotated.2022-03-28_200360.zip with unzip -o
> Archive:  /sgm/queue//4//14322498/upload/19829_2022-04-05T13-10-06.577_ctt-data-rotated.2022-03-28_200360.zip
>   inflating: data/rotated/CTT-0825A4690059-gps.2022-03-25_182041.csv.gz  
>   inflating: data/rotated/CTT-0825A4690059-gps.2022-03-25_192041.csv.gz  
> ...
>   inflating: data/rotated/CTT-0825A4690059-raw-data.2022-03-28_173235.csv.gz  
>   inflating: data/rotated/CTT-0825A4690059-raw-data.2022-03-28_183236.csv.gz  
>   inflating: data/rotated/CTT-0825A4690059-raw-data.2022-03-28_193237.csv.gz  
>

**Comment by denislepage** [comment]
The log confirms that 215+ ctt files were pushed up to AWS around that time for this particular receiver, so we’ll need to track when the data might be coming back.

**Comment by cjardine-bsc** [comment]
I believe this is now resolved. @denislepage @j-sayers @leberrigan  Close?
